public/js/components/specs/app.spec.js

```diff
@@ -0,0 +1,243 @@
+
+    title[1] = "\u2700";
+    title[2] = "IoT Dashboard";
+
+    expect(vm.appTitle).toEqual(
+      `<span class="hemoji">${title[1]}</span> ${title[2]}`
+    );
+  });
+
+  test("onSaveSettings method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.dashboard).toEqual(
+      expect.objectContaining({
+        bgColor: expect.any(String),
+        bgImageRepeat: expect.any(Boolean),
+        bgImageUrl: expect.any(String),
+        blockSize: expect.any(Array),
+        editMode: expect.any(String),
+        tiles: expect.any(Array),
+        title: expect.any(String)
+      })
+    );
+
+    vm.onSaveSettings(event => {
+      event = {
+        bgColor: "#fff",
+        bgImageRepeat: true,
+        bgImageUrl: "",
+        title: "\u2700 IoT Dashboard"
+      };
+      expect(vm.saveDashboard).toHaveBeenCalled();
+    });
+
+    Vue.nextTick(() => {
+      expect(vm.dashboard).toEqual(
+        expect.objectContaining({
+          bgColor: expect.any(String),
+          bgImageRepeat: expect.any(Boolean),
+          bgImageUrl: expect.any(String),
+          blockSize: expect.any(Array),
+          editMode: expect.any(String),
+          tiles: expect.any(Array),
+          title: expect.any(String)
+        })
+      );
+    });
+  });
+
+  test("onTileChange method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.dashboard).toEqual(mockDashboardData.dashboard);
+
+    mockDashboardData.dashboard.tiles[0].buttonText = "stop it";
+
+    vm.onTileChange(event => {
+      event.id = mockDashboardData.dashboard.tiles[0].id;
+      expect(vm.saveDashboard).toHaveBeenCalled();
+    });
+
+    expect(vm.dashboard.tiles[0]).toEqual({
+      buttonText: "stop it",
+      deviceId: "AZ3166",
+      deviceMethod: "stop",
+      id: "2471d5ab-0d73-42a3-ba4f-f694574feb6b",
+      position: [54, 466],
+      size: [0.8, 0.7],
+      title: "MXChip sending",
+      type: "button"
+    });
+  });
+
+  test("onTileDelete method", () => {
+    const wrapper = shallowMountApp();
+    const tileId = mockDashboardData.dashboard.tiles[0].id;
+
+    wrapper.vm.onTileDelete(tileId);
+
+    expect(
+      wrapper.find({ name: "base-card" }).vm.$emit("tile-delete")
+    ).toBeTruthy();
+
+    expect(wrapper.vm.dashboard.tiles.length).toBe(1);
+  });
+
+  test("onTileCreate method", () => {
+    const { vm } = shallowMountApp();
+
+    vm.onTileCreate(event => {
+      event = {
+        buttonText: "begin",
+        deviceId: "DH7643",
+        deviceMethod: "halt",
+        id: "8071d5ab-0z12-41i3-ba9p-f600124feb6b",
+        position: [106, 266],
+        size: [0.9, 0.3],
+        title: "MSchip receiving",
+        type: "button"
+      };
+    });
+
+    expect(vm.dashboard.tiles.length).toBe(3);
+  });
+
+  test("onDeviceListRecieved method", () => {
+    const { vm } = shallowMountApp();
+    const io = jest.fn();
+    const socket = io();
+
+    vm.onDeviceListReceived(() => {
+      expect(socket).toHaveBeenCalled();
```

Comment:
Couple of things here: the `onDeviceListReceived` is called inside the `created` hook of the Vue component, so we could get a bit closer to the actual behaviour of the component in the test by not calling `vm.onDeviceListReceived` directly, and instead mounting the component and asserting that `socket` was called.

Secondly, this `expect` is not reached by the test, as was the case with the two tests I commented on above. We should be able to run it like this, but it fails:

```javascript
vm.onDeviceListReceived();
expect(socket).toHaveBeenCalled();
```

I think it might be more useful to spy on the contents of the `socket.on` callback than to check if the `const socket = io();` ran correctly, We could do this with [this kind of approach](https://stackoverflow.com/questions/46204810/how-to-test-socket-io-events-being-called-with-sinon-spy) (it references `sinon.spy` but Jest can do something similar).

Reply:
Ok gotcha - thanks for the detailed feedback here! Looking at the the mentioned approach now.

Reply:
Ok, so, I've been having some difficulty writing a similar test that you mentioned [here](https://stackoverflow.com/questions/46204810/how-to-test-socket-io-events-being-called-with-sinon-spy). I updated the pr based upon all your feedback but left this as a TODO item. 

Just as reference, this is where I left off from the test: 
```
  test("onDeviceListReceived method", () => {
    const { vm } = shallowMountApp();
    const message = jest.fn();

    io().on("message", message);

    expect(message).toHaveBeenCalledWith(vm.onDeviceListReceived());
  });
```

and the error I was getting (jest mock function was never called): 
![Screen Shot 2019-06-10 at 10 59 31 PM](https://user-images.githubusercontent.com/24294676/59242986-75123a80-8bd3-11e9-8d9e-070ce5c1d944.png)

I tried a few other variations but they all led to this error. Hopefully I'm missing something simple. I've never written a test for something like this before so it's a bit new to me 😅 (which is awesome).

But any who, I appreciate all the feedback and look forward to any additional feedback you have on this 😊

Reply:
To be honest, looking at this again, I'm not even sure I suggested the _right_ thing to test. 😄 We can assume that the `socket.io` library is thoroughly tested, so checking that `socket.on` was called probably isn't so useful for us.

What would be useful is to check that when `"message"` is received, `this.messages` is updated with the right information. So we could look into mocking socket.io events (maybe with a library [like `mock-socket`](https://github.com/thoov/mock-socket)) and checking the outcome, i.e. that `this.messages` is updated with the correct `message.body`. We could also check that the `if (this.messages.length > 500)` conditional behaves as expected.

However, I think this test could be a PR all on its own, so it shouldn't be a blocker for merging this good work. 👍

Reply:
Ah, I see what you're saying. Well, I'll definitely be creating another pr soon and try and knock this one out 😅